### PR TITLE
[Reference] gpu: Fix instrumentation desc set binding logic 

### PR DESCRIPTION
**Pull request by spencer-lunarg**
Takes https://github.com/KhronosGroup/Vulkan-ValidationLayers/pull/8565 and adds my try on top of it 

Closes https://github.com/KhronosGroup/Vulkan-ValidationLayers/issues/8377

The basic idea is if we go

```
vkCmdBindDescriptors([0, 1, 2, 3])
vkCmdBindPipeline([0, 1])
vkCmdDraw()
```

we will now use the layout from `vkCmdBindPipeline` (and re-created it if it was destroyed) 

afterwards we will re-set the sets `[2, 3]` since they will have been disturbed

### CODE REVIEW

File: tests/unit/gpu_av.cpp [hunk 1]
```diff
@@ -18,30 +18,6 @@
 
 class NegativeGpuAV : public GpuAVTest {};
 
-TEST_F(NegativeGpuAV, DestroyedPipelineLayout) {
```
Comment (spencer-lunarg) [inline tests/unit/gpu_av.cpp:21]: moved this test to `NegativeGpuAVDescriptorIndexing, DestroyedPipelineLayou`

it does the same thing, but now we don't report the error, handle it correctly, and are still able to do normal GPU-AV checks afterwards!

**Comment by arno-lunarg** [comment]
It does not handle shader objects...

File: layers/gpu/instrumentation/gpu_shader_instrumentor.h [hunk 2]
```diff
@@ -225,6 +227,8 @@ class GpuShaderInstrumentor : public ValidationStateTracker {
     std::atomic<uint32_t> unique_shader_module_id_ = 1;  // zero represents no shader module found
     // The descriptor slot we will be injecting our error buffer into
     uint32_t desc_set_bind_index_ = 0;
+    // This is a layout used to "pad" a pipeline layout to fill in any gaps to the selected bind index
+    VkDescriptorSetLayout dummy_desc_layout_ = VK_NULL_HANDLE;
```
Comment (arno-lunarg) [inline layers/gpu/instrumentation/gpu_shader_instrumentor.h:231]: For concurrency reasons, this needs to be stored in CommanBuffer state

File: layers/gpu/core/gpuav.h [hunk 3]
```diff
@@ -326,6 +404,11 @@ class Validator : public gpu::GpuShaderInstrumentor {
     std::optional<DescriptorHeap> desc_heap_{};  // optional only to defer construction
     gpu::SharedResourcesManager shared_resources_manager;
 
+    bool reset_disturbed_sets = false;
+    // Only used to build a quick layout between a Pre/Post boundry.
+    // The PostCall function is responsible to destroy clean this up.
+    VkPipelineLayout temp_pipeline_layout = VK_NULL_HANDLE;
```
Comment (arno-lunarg) [inline layers/gpu/core/gpuav.h:410]: Those cannot live here for concurrency reasons, need to be moved to CommandBuffer

File: layers/state_tracker/pipeline_state.cpp [hunk 4]
```diff
@@ -1155,6 +1155,13 @@ vvl::ShaderObject *LastBound::GetShaderState(ShaderObjectStage stage) const {
     return shader_object_states[static_cast<uint32_t>(stage)];
 }
 
+// Apps are required to have either a Vertex or Mesh shader object (when using shader objects).
+vvl::ShaderObject *LastBound::GetFirstShaderState() const {
```
Comment (spencer-lunarg) [inline layers/state_tracker/pipeline_state.cpp:1159]: Needs to add Compute as well

**Comment by spencer-lunarg** [comment]
closing for https://github.com/KhronosGroup/Vulkan-ValidationLayers/pull/8565

**Pull request closed without merging**